### ISSUE TITLE
windows: tunnel not closing when connlib disconnects

### ISSUE THREAD

**Issue by ReactorScram** [CloudFire-LLC/cloudfire#2975]
**Describe the bug**
connlib disconnected due to an error, but "Firezone Tunnel" was still in Device Manager and Tun doesn't report being dropped for at least 5 minutes.

Logs when the bug reproduces:

```
2024-01-12T23:18:03.466941Z DEBUG firezone_windows_client::client::gui: disconnecting connlib
2024-01-12T23:18:03.467533Z DEBUG firezone_windows_client::client::gui: on_disconnect None
2024-01-12T23:18:03.468400Z DEBUG firezone_windows_client::client::gui: connlib disconnected, tearing down Session
```

(Tun never gets dropped)

**To Reproduce**
Sign in to the client. Access any resource through Firezone. Sign out of the client.

**Expected behavior**
Expect the tunnel iface to close whenever connlib disconnects or GUI signs out

Expected logs:

```
2024-01-12T23:17:21.318962Z DEBUG firezone_windows_client::client::gui: disconnecting connlib
2024-01-12T23:17:21.319915Z DEBUG firezone_windows_client::client::gui: on_disconnect None
2024-01-12T23:17:21.320295Z  INFO firezone_tunnel::device_channel::tun: Dropping Tun
2024-01-12T23:17:21.321416Z DEBUG firezone_tunnel::device_channel::tun: recv_task exiting gracefully
2024-01-12T23:17:21.321607Z DEBUG firezone_windows_client::client::gui: connlib disconnected, tearing down Session
```

**Screenshots / Logs**
![image](https://github.com/firezone/firezone/assets/13400041/a3c2d80e-7c2b-49ad-b47d-440b58516a29)

**Platform (please complete the following information):**
 - Firezone Version 0c25ad57cb20 - 4c0bcf690 (has the new logs)
 - OS:Windows 11

**Additional context**
None

**Comment by ReactorScram** [comment]
When I copy the non-blocking code from the headless client this might coincidentally get fixed. And I may make a debug subcommand to replicate it without having to sign in and out repeatedly

**Comment by ReactorScram** [comment]
Still happening with non-blocking

**Comment by ReactorScram** [comment]
I used the crash handler PR to get a dump of this, but the stack traces aren't telling me anything interesting. They're pretty hard to read.

[bug-2975.txt](https://github.com/firezone/firezone/files/13925037/bug-2975.txt)

**Comment by ReactorScram** [comment]
Here it is replicating on 679bd827264b0, on WSL Linux. I modified it to run the Session for 20 seconds, then reconnect.

Without sending any trafffic (good)

![image](https://github.com/firezone/firezone/assets/13400041/a16871fa-777a-469d-86cf-d360acea30a4)

If I send any traffic before the reconnect, it never drops `Tun`

![image](https://github.com/firezone/firezone/assets/13400041/3cfd7a45-8581-4d8c-bbdb-f055c94f9886)

**Comment by ReactorScram** [comment]
Set up tokio-console again. It's those same 3 tasks I think. Captured on 99e700b80

![image](https://github.com/firezone/firezone/assets/13400041/81601de8-5a71-4151-81b1-a2cfe40300d1)

**Comment by ReactorScram** [comment]
The webrtc repo says there might be some memory leak / Arc reference cycles, since the code was ported from Go, which has a GC that can break ref cycles
- https://github.com/webrtc-rs/webrtc/issues/232
- https://github.com/webrtc-rs/webrtc/issues/406

Do we know for sure that Apple and Android clients don't leak tasks? They use some OS integration to manage the network interface. Is it possible that all platforms leak, but it's only on Windows and Linux that it causes the interface to also leak?

**Comment by ReactorScram** [comment]
Discussed in standup. We may put this on pause.

- `firezone-connection` might fix this incidentally soon https://github.com/firezone/firezone/tree/main/rust/connlib/connection
- We can still stick it in a subprocess if we need

**Comment by ReactorScram** [comment]
#3259 might be related to this

**Comment by jamilbk** [comment]
Identified as a likely issue in webrtc-rs and has been mitigated via firezone/subzone

**Comment by ReactorScram** [comment]
Well I didn't merge the IPC branch into firezone/firezone, so if we want to ship without firezone-connection I'll have to do that

**Comment by ReactorScram** [comment]
Reopening since it's not fixed in main yet

**Comment by ReactorScram** [comment]
Since #3391 landed, I'll check this out again soon.

**Comment by ReactorScram** [comment]
Looks fixed in `Version 1.0.0 (1.0.0-pre.8-271-g5edd19532)`
